File: golang/internal/db/query.go
package db

import (
	"context"
	"fmt"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.opentelemetry.io/otel/codes"
	"inetum.com/metrics-go-app/internal/otel"
)

func ReadData(ctx context.Context, collection *mongo.Collection, opts bson.M) ([]Data, error) {
	ctx, span := otel.Tracer.Start(ctx, "ReadData")
	defer span.End()
	span.AddEvent("ReadData started")

	findOutput, err := collection.Find(ctx, opts)
	if err != nil {
		fmt.Println(err)
		span.SetStatus(codes.Error, err.Error())
		span.RecordError(err)
		return nil, err
	}

	var data []Data
	if err := findOutput.All(ctx, &data); err != nil {
		fmt.Println(err)
		span.SetStatus(codes.Error, err.Error())
		span.RecordError(err)
		return nil, err
	}

	span.SetStatus(codes.Ok, "ok")
	span.AddEvent("ReadData done")
	return data, nil
}

File: dotnet/Database/Query.cs
using System.Diagnostics;
using MongoDB.Bson;
using MongoDB.Driver;
using OpenTelemetry.Trace;

namespace DotnetAppBlueprint;

public record DataDto(int Data);

internal static class Query
{
    public static async Task<List<DataDto>> ReadData(
        this IMongoCollection<BsonDocument> collection,
        FilterDefinition<BsonDocument> filter,
        Instrumentation instrumentation,
        ILogger logger)
    {
        using var activity = instrumentation.ActivitySource.StartActivity("ReadData")!;
        activity.AddEvent(new("ReadData started"));

        try
        {
            var result = await collection.Find(filter).ToListAsync();

            activity.SetStatus(ActivityStatusCode.Ok, "ok");
            activity.AddEvent(new("ReadData done"));
            return result.Select(r => new DataDto(r.GetValue("data").AsInt32)).ToList();
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "{Message}", ex.Message);

            activity.SetStatus(ActivityStatusCode.Error, "UnexpectedError");
            activity.RecordException(ex);
            throw;
        }
    }
}
